Test on model producing NULL date values fail with empty diff

# Issue description

Let's say we have an external model `model1` defined as follows:
```yaml
- name: dummy.model1
  columns:
    id: INT
    date: DATE
```
And a model `model2` using `model1` defined as follows:
```sql
MODEL (
  name dummy.model2,
  kind FULL,
);

SELECT
  *,
  DATE_TRUNC('month', date)::DATE AS month,
  NULL::DATE AS null_date,
FROM
  dummy.model1
```
And finally a unit test for `model2` defined as follows:
```yaml
test_null_dates:
  model: dummy.model2
  inputs:
    dummy.model1:
      rows:
        - id: 1234
          date: 2023-01-12
        - id: 9876
          date: 2023-02-10
  outputs:
    query:
      rows:
        - id: 1234
          date: 2023-01-12
          month: 2023-01-01
          null_date:
        - id: 9876
          date: 2023-02-10
          month: 2023-02-01
          null_date:
```

When running this test, I get the following error:
```bash
======================================================================
FAIL: test_null_dates (tests/test_model2.yaml)
----------------------------------------------------------------------
AssertionError: Data differs (exp: expected, act: actual)

Empty DataFrame
Columns: []
Index: []

----------------------------------------------------------------------
Ran 1 test in 0.036s

FAILED (failures=1)
```

# Analysis

In the `ModelTest.assert_equal()` method the issue seems to be related to [this line](https://github.com/TobikoData/sqlmesh/blob/573951086edf1c51366e8783e49dfb60fe24e914/sqlmesh/core/test/definition.py#L104).
This line updates the `actual` dataframe types, replace the `DATE` columns schemas.

Before this line:
```
actual.dtypes=('id', dtype('int32')) ('date', dtype('<M8[us]')) ('month', dtype('<M8[us]')) ('null_date', dtype('<M8[us]'))
```

After this lines:
```
actual.dtypes=('id', dtype('int32')) ('date', dtype('O')) ('month', dtype('O')) ('null_date', dtype('O'))
```

@georgesittas that's strange because I still have this issue on a new clean environment created from the `main` branch:
```bash
$ poetry init

$ poetry add git+https://github.com/TobikoData/sqlmesh.git
...
  • Installing sqlglot (19.8.2)
  • Installing sqlmesh (0.57.2.dev18 5739510)

$ poetry run python -V
Python 3.11.2

$ poetry run sqlmesh --version
0.57.2.dev18

$ poetry run sqlmesh init duckdb
```
```bash
$ cat schema.yaml 
- name: dummy.model1
  columns:
    id: INT
    date: DATE
```
```bash
$ cat models/model2.sql 
MODEL (
  name dummy.model2,
  kind FULL,
);

SELECT
  *,
  DATE_TRUNC('month', date)::DATE AS month,
  NULL::DATE AS null_date,
FROM
  dummy.model1
```
```bash
$ cat tests/test_model2.yaml 
test_null_dates:
  model: dummy.model2
  inputs:
    dummy.model1:
      rows:
        - id: 1234
          date: 2023-01-12
        - id: 9876
          date: 2023-02-10
  outputs:
    query:
      rows:
        - id: 1234
          date: 2023-01-12
          month: 2023-01-01
          null_date:
        - id: 9876
          date: 2023-02-10
          month: 2023-02-01
          null_date:
```
```bash
$ poetry run sqlmesh test tests/test_model2.yaml 
F
======================================================================
FAIL: test_null_dates (tests/test_model2.yaml)
----------------------------------------------------------------------
AssertionError: Data differs (exp: expected, act: actual)

Empty DataFrame
Columns: []
Index: []

----------------------------------------------------------------------
Ran 1 test in 0.029s

FAILED (failures=1)
```

The issue may come from my environment but I'm a bit lost as I can reproduce this in a newly created virtual environment.

Hey @pcasteran, I looked into this again and I found that this issue occurs when using DuckDB v0.9.2. In particular, the difference between the expected and the actual dataframe seems to be:

```python
-> pd.testing.assert_frame_equal(
(Pdb) expected
     id       date      month null_date
0  1234 2023-01-12 2023-01-01      None
1  9876 2023-02-10 2023-02-01      None
(Pdb) actual
     id                 date                month null_date
0  1234  2023-01-12 00:00:00  2023-01-01 00:00:00      None
1  9876  2023-02-10 00:00:00  2023-02-01 00:00:00      None
(Pdb) pd.testing.assert_frame_equal(expected.sort_index(axis=1),actual.sort_index(axis=1),check_dtype=False,check_datetimelike_compat=True)
*** AssertionError: [datetimelike_compat=True] <DatetimeArray>
['2023-01-12 00:00:00', '2023-02-10 00:00:00']
Length: 2, dtype: datetime64[us] is not equal to [Timestamp('2023-01-12 00:00:00') Timestamp('2023-02-10 00:00:00')].
(Pdb)
```

So, it seems like there are some type shenanigans going on here... I'll need to investigate a bit more to understand this better, but at least now we know what's up. Appreciate your patience on this, will let you know when I have an update. 

<details>
<summary>Click this to see details on how to reproduce in CLI</summary>

```
➜  tree .
.
├── audits
├── config.yaml
├── macros
├── models
│   └── model2.sql
├── schema.yaml
├── seeds
└── tests
    └── test_model2.yaml

6 directories, 4 files
➜  cat models/model2.sql schema.yaml tests/test_model2.yaml
MODEL (
  name dummy.model2,
  kind FULL,
);

SELECT
  *,
  DATE_TRUNC('month', date)::DATE AS month,
  NULL::DATE AS null_date,
FROM
  dummy.model1
- name: dummy.model1
  columns:
    id: INT
    date: DATE
test_null_dates:
  model: dummy.model2
  inputs:
    dummy.model1:
      rows:
        - id: 1234
          date: 2023-01-12
        - id: 9876
          date: 2023-02-10
  outputs:
    query:
      rows:
        - id: 1234
          date: 2023-01-12
          month: 2023-01-01
          null_date:
        - id: 9876
          date: 2023-02-10
          month: 2023-02-01
          null_date:
➜  pip show duckdb
Name: duckdb
Version: 0.9.2
...
➜  sqlmesh test
F
======================================================================
FAIL: test_null_dates (tests/test_model2.yaml)
----------------------------------------------------------------------
AssertionError: Data differs (exp: expected, act: actual)

Empty DataFrame
Columns: []
Index: []

----------------------------------------------------------------------
Ran 1 test in 0.010s

FAILED (failures=1)
➜  pip install "duckdb==0.9.1" --upgrade
...
Successfully installed duckdb-0.9.1
➜  pip show duckdb
Name: duckdb
Version: 0.9.1
...
➜  sqlmesh test
.
----------------------------------------------------------------------
Ran 1 test in 0.011s

OK
```

</details>

--- a/sqlmesh/core/test/definition.py
+++ b/sqlmesh/core/test/definition.py
@@ -100,8 +100,8 @@ class ModelTest(unittest.TestCase):
             actual_types, errors="ignore"
         )
 
-        expected = expected.replace({np.nan: None, "nan": None})
-        actual = actual.replace({np.nan: None, "nan": None})
+        expected = expected.replace({None: np.nan, "nan": np.nan})
+        actual = actual.replace({None: np.nan, "nan": np.nan})
 
         try:
             pd.testing.assert_frame_equal(

--- a/sqlmesh/core/test/definition.py
+++ b/sqlmesh/core/test/definition.py
@@ -100,8 +100,8 @@ class ModelTest(unittest.TestCase):
             actual_types, errors="ignore"
         )
 
-        expected = expected.replace({None: np.nan, "nan": np.nan})
-        actual = actual.replace({None: np.nan, "nan": np.nan})
+        expected = expected.replace({None: np.nan})
+        actual = actual.replace({None: np.nan})
 
         try:
             pd.testing.assert_frame_equal(

--- a/tests/core/test_test.py
+++ b/tests/core/test_test.py
@@ -222,34 +222,9 @@ test_foo:
     assert result and result.wasSuccessful()
 
 
-def test_nan(sushi_context: Context, full_model_without_ctes: SqlModel) -> None:
-    model = t.cast(SqlModel, sushi_context.upsert_model(full_model_without_ctes))
-    body = load_yaml(
-        """
-test_foo:
-  model: sushi.foo
-  inputs:
-    raw:
-      - id: 1
-        value: nan
-        ds: 3
-  outputs:
-    query:
-      - id: 1
-        value: null
-        ds: 3
-  vars:
-    start: 2022-01-01
-    end: 2022-01-01
-        """
-    )
-    result = _create_test(body, "test_foo", model, sushi_context).run()
-    assert result and result.wasSuccessful()
-
-
 def test_partial_data(sushi_context: Context) -> None:
     model = _create_model(
-        "WITH source AS (SELECT id, name FROM sushi.waiter_names) SELECT id, name FROM source"
+        "WITH source AS (SELECT id, name FROM sushi.waiter_names) SELECT id, name, 'nan' as str FROM source"
     )
     model = t.cast(SqlModel, sushi_context.upsert_model(model))
 
@@ -274,9 +249,12 @@ test_foo:
           name: 'bob'
     query:
       - id: 1
+        str: nan
       - id: 2
+        str: nan
       - id: 3
         name: 'bob'
+        str: nan
         """
     )
     result = _create_test(body, "test_foo", model, sushi_context).run()

--- a/tests/core/test_test.py
+++ b/tests/core/test_test.py
@@ -334,6 +334,36 @@ test_foo:
     result = _create_test(body, "test_foo", model, sushi_context).run()
     assert result and result.wasSuccessful()
 
+    model = _create_model(
+        "SELECT *, DATE_TRUNC('month', date)::DATE AS month, NULL::DATE AS null_date, FROM unknown"
+    )
+    model = t.cast(SqlModel, sushi_context.upsert_model(model))
+
+    body = load_yaml(
+        """
+test_foo:
+  model: sushi.foo
+  inputs:
+    unknown:
+      - id: 1234
+        date: 2023-01-12
+      - id: 9876
+        date: 2023-02-10
+  outputs:
+    query:
+      - id: 1234
+        date: 2023-01-12
+        month: 2023-01-01
+        null_date:
+      - id: 9876
+        date: 2023-02-10
+        month: 2023-02-01
+        null_date:
+        """
+    )
+    result = _create_test(body, "test_foo", model, sushi_context).run()
+    assert result and result.wasSuccessful()
+
 
 def test_missing_column_failure(sushi_context: Context, full_model_without_ctes: SqlModel) -> None:
     model = t.cast(SqlModel, sushi_context.upsert_model(full_model_without_ctes))
@@ -355,7 +385,7 @@ test_foo:
     result = _create_test(body, "test_foo", model, sushi_context).run()
     assert result and not result.wasSuccessful()
 
-    expected_msg = "AssertionError: Data differs (exp: expected, act: actual)\n\n  value        ds    \n    exp act   exp act\n0  None   2  None   3\n"
+    expected_msg = "AssertionError: Data differs (exp: expected, act: actual)\n\n  value      ds    \n    exp act exp act\n0   NaN   2 NaN   3\n"
     assert expected_msg in result.failures[0][1]
 
 

